File: app/src/main/java/net/veldor/flibustaloader/utils/Grammar.java
package net.veldor.flibustaloader.utils;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.text.Html;
import android.util.Log;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.selections.Author;

import java.util.Random;

public class Grammar {
    @SuppressWarnings("deprecation")
    public static String textFromHtml(String html) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return (Html.fromHtml(html, Html.FROM_HTML_MODE_COMPACT)).toString();
        } else {
            return (Html.fromHtml(html)).toString();
        }
    }

    public static int getRandom() {
        int min = 1000;
        int max = 9999;

        Random r = new Random();
        return r.nextInt(max - min + 1) + min;
    }

    public static String getAppVersion() {
        try {
            PackageInfo pInfo = App.getInstance().getPackageManager().getPackageInfo(App.getInstance().getPackageName(), 0);
            return pInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            Log.d("surprise", "MainActivity setupUI: can't found version");
            e.printStackTrace();
        }
        return "0";
    }

    public static String createAuthorDirName(Author author) {
        String dirname;
        dirname = author.name;
       /* // разобью имя автора по пробелу
        String[] parts = author.name.split(" ");
        if (parts.length > 1) {
            dirname = parts[0] + ' ' + parts[1];
        }
        else {
            dirname = parts[0];
        }*/
        if (dirname.length() > 100) {
            return dirname.substring(0, 100);
        }
        return dirname;
    }

    public static String clearDirName(String dirName) {
        return dirName.replaceAll("[^ а-яА-Яa-zA-Z0-9.\\-]", "");
    }

    public static String getExtension(String filename){
        return filename.substring(filename.lastIndexOf(".") + 1);
    }

    public static String changeExtension(String name, String trueFormatExtension) {
        return name.substring(0, name.lastIndexOf(".")) + "." + trueFormatExtension;
    }

    public static String getLiteralSize(long length) {
        if(length > 1024 * 1024){
            return length / 1024 / 1024 + " мб.";
        }
        if(length > 1024){
            return length / 1024 + " кб.";
        }
        return length + " байт";
    }
}

File: app/src/main/java/net/veldor/flibustaloader/utils/Grammar.kt
package net.veldor.flibustaloader.utils

import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.text.Html
import android.util.Log
import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.selections.Author
import net.veldor.flibustaloader.selections.FoundedEntity
import java.util.*
import android.text.Spannable

import android.text.style.ForegroundColorSpan

import android.text.SpannableString
import android.text.style.BackgroundColorSpan
import android.widget.TextView


object Grammar {
    @kotlin.jvm.JvmStatic
    fun textFromHtml(html: String?): String {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Html.fromHtml(html, Html.FROM_HTML_MODE_COMPACT).toString()
        } else {
            Html.fromHtml(html).toString()
        }
    }

    @kotlin.jvm.JvmStatic
    val random: Int
        get() {
            val min = 1000
            val max = 9999
            val r = Random()
            return r.nextInt(max - min + 1) + min
        }

    @kotlin.jvm.JvmStatic
    val appVersion: String
        get() {
            try {
                val pInfo = App.instance.packageManager.getPackageInfo(
                    App.instance.packageName,
                    0
                )
                return pInfo.versionName
            } catch (e: PackageManager.NameNotFoundException) {
                Log.d("surprise", "MainActivity setupUI: can't found version")
                e.printStackTrace()
            }
            return "0"
        }

    @kotlin.jvm.JvmStatic
    fun createAuthorDirName(author: Author): String {
        val dirname: String = author.name!!
        return if (dirname.length > 100) {
            dirname.substring(0, 100)
        } else dirname
    }

    @kotlin.jvm.JvmStatic
    fun createAuthorDirName(author: FoundedEntity): String {
        val dirname: String = author.name!!
        return if (dirname.length > 100) {
            dirname.substring(0, 100).trim()
        } else dirname.trim()
    }

    @kotlin.jvm.JvmStatic
    fun clearDirName(dirName: String): String {
        return Regex("[^\\d\\w ]").replace(dirName, "")
    }

    @kotlin.jvm.JvmStatic
    fun getExtension(filename: String): String {
        return filename.substring(filename.lastIndexOf(".") + 1)
    }

    fun changeExtension(name: String, trueFormatExtension: String?): String {
        return name.substring(0, name.lastIndexOf(".")) + "." + trueFormatExtension
    }

    @kotlin.jvm.JvmStatic
    fun getLiteralSize(length: Long): String {
        if (length > 1024 * 1024) {
            return (length / 1024 / 1024).toString() + " мб."
        }
        return if (length > 1024) {
            (length / 1024).toString() + " кб."
        } else "$length байт"
    }

    fun removeExtension(name: String): String {
        return name.substring(0, name.lastIndexOf("."))
    }

    fun isValidUrl(newValue: String): Boolean {
        return "^https?://\\w+.\\w+\$".toRegex().matches(newValue)
    }

    fun getAvailableDownloadFormats(item: FoundedEntity, view: TextView) {
        view.text = ""
        if (item.downloadLinks.isEmpty()) {
            view.text = getColoredString("Не найдены ссылки для загрузки", Color.RED)
        }
        item.downloadLinks.forEach {
            if (it.mime != null) {
                if (it.mime!!.contains("fb2")) {
                    view.append(getColoredString(" FB2 ", Color.parseColor("#FFE91E63")))
                    view.append(" ")
                } else if (it.mime!!.contains("mobi")) {
                    view.append(getColoredString(" MOBI ", Color.parseColor("#FF9C27B0")))
                    view.append(" ")
                } else if (it.mime!!.contains("epub")) {
                    view.append(getColoredString(" EPUB ", Color.parseColor("#FF673AB7")))
                    view.append(" ")
                } else if (it.mime!!.contains("pdf")) {
                    view.append(getColoredString(" PDF ", Color.parseColor("#FF3F51B5")))
                    view.append(" ")
                } else if (it.mime!!.contains("txt")) {
                    view.append(getColoredString(" TXT ", Color.parseColor("#FF2196F3")))
                    view.append(" ")
                } else if (it.mime!!.contains("html")) {
                    view.append(getColoredString(" HTML ", Color.parseColor("#FF009688")))
                    view.append(" ")
                } else if (it.mime!!.contains("doc")) {
                    view.append(getColoredString(" DOC ", Color.parseColor("#FF4CAF50")))
                    view.append(" ")
                } else if (it.mime!!.contains("djvu")) {
                    view.append(getColoredString(" DJVU ", Color.parseColor("#FFFF9800")))
                    view.append(" ")
                } else if (it.mime!!.contains("rtf")) {
                    view.append(getColoredString(" RTF ", Color.parseColor("#030303")))
                    view.append(" ")
                } else {
                    view.append(getColoredString(it.mime!!, Color.parseColor("#030303")))
                    view.append(" ")
                }
            }
        }
    }

    fun getColoredString(mString: String?, colorId: Int): Spannable {
        val spannable: Spannable = SpannableString(mString)
        spannable.setSpan(
            BackgroundColorSpan(colorId),
            0,
            spannable.length,
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        return spannable
    }
}

